### DISCUSSION TITLE
Nominatim docker stuck on "database system is ready to accept connections"

### DISCUSSION THREAD

**Discussion by cormanttech** [mediagis/nominatim-docker#339] [Q&A] [answered]
Hi All,

When I run the docker-compose.yml, the loading stuck at "database system is ready to accept connections"

![image](https://user-images.githubusercontent.com/55179774/177492457-ffa9ecf5-3cb6-4617-a327-40e066661252.png)

Port 5432 is free on my server:

![image](https://user-images.githubusercontent.com/55179774/177492593-4236e9a9-e5d3-49b1-8ba4-2a6edd405bf1.png)

Is there any configuration that I miss? Thank you.

**Comment by philipkozeny** [comment, marked as answer]
It's not stuck, it's done with the import & setup :)

**Reply by cormanttech** [reply]
Thanks for this sir!

But, why is that when I am checking port 5432, nothing is using it? And also, when I am accessing the localhost/, I am getting this 

![image](https://user-images.githubusercontent.com/55179774/177494307-ed88be51-4b2d-48ff-a78c-bc38f6956412.png)

**Reply by mtmail** [reply]
You said you use the `docker-compose.yml` file. Check in the file which ports you selected. https://docs.docker.com/compose/compose-file/#ports

**Reply by cormanttech** [reply]
Hi Sir,

The configured port in docker-compose.yml is 8080:8080

![image](https://user-images.githubusercontent.com/55179774/177679849-246e9fe1-e83c-4960-918d-2baa255ac305.png)

But when I try to access it on web browser localhost:8080, this is what it shown:

![image](https://user-images.githubusercontent.com/55179774/177680825-cb96fdde-04ea-4e26-aa7a-3f789bb97cc9.png)

**Reply by philipkozeny** [reply]
This means, that everything is setup and works. You send a request to the API without a search parameter, so the result is empty ('[]'). Try to search for something specific like we have described in the docs: https://github.com/mediagis/nominatim-docker/blob/master/4.0/README.md#automatic-import so something like http://localhost:8080/search.php?q=manila

**Reply by cormanttech** [reply]
Hi Sir Philip,

This is well noted. So does this mean that the Nominatim Container that I cloned doesn't have a UI? https://github.com/mediagis/nominatim-docker/blob/master/4.0/README.md#automatic-import 

Thanks sir!

**Reply by cormanttech** [reply]
I tried this link on my web browser  http://localhost:8080/search.php?q=manila and I got this result:

![image](https://user-images.githubusercontent.com/55179774/178167690-8e2811d5-7db8-4856-9c36-336fcaca0bb0.png)

Is this the correct result?

**Reply by philipkozeny** [reply]
> I tried this link on my web browser http://localhost:8080/search.php?q=manila and I got this result:
> 
> ![image](https://user-images.githubusercontent.com/55179774/178167690-8e2811d5-7db8-4856-9c36-336fcaca0bb0.png)
> 
> Is this the correct result?

Yes, thats the correct result & yes, the UI was moved from that the upstream nominatim project in another repository.

**Reply by cormanttech** [reply]
Hi Sir @philipkozeny 


May I ask if it is possible to connect this nominatim-docker into this Nominatim UI https://github.com/osm-search/nominatim-ui/

Thank you very much!

**Reply by mtmail** [reply]
This task requires a basic understanding of Linux system administration, Docker, network ports and JSON.

I belive https://github.com/osm-search/nominatim-ui/ contains enough instructions to start the nominatim-ui app and how to configure it to point to a Nominatim API endpoint. The documentation links to 20+ different projects how to start a static webserver. Search for ["docker static webserver"]( https://duckduckgo.com/?t=ffab&q=docker+static+webserver&ia=web) if you need a static webserver inside a docker container.

**Reply by cormanttech** [reply]
Hi Sir,

I already connected the Nominatim API to Nominatim UI and I can now access and use it using localhost:8765 using Phyton.

My, hopefully, the last problem now is forwarding it to Apache. I follow the instructions here https://nominatim.org/release-docs/latest/admin/Setup-Nominatim-UI/ for Setting up forwarding in Apache.

I include this in /etc/Apache2/conf-available/nominatim.conf

<Directory "/srv/nominatim/nominatim-project/website">
  Options FollowSymLinks MultiViews
  AddType text/html   .php
  Require all granted

  RewriteEngine On

  # This must correspond to the URL where nominatim can be found.
  RewriteBase "/nominatim/"

  # If no endpoint is given, then use search.
  RewriteRule ^(/|$)   "search.php"

  # If format-html is explicity requested, forward to the UI.
  RewriteCond %{QUERY_STRING} "format=html"
  RewriteRule ^([^/]+).php ui/$1.html [R,END]
  # Same but .php suffix is missing.
  RewriteCond %{QUERY_STRING} "format=html"
  RewriteRule ^([^/]+) ui/$1.html [R,END]

  # If no format parameter is there then forward anything
  # but /reverse and /lookup to the UI.
  RewriteCond %{QUERY_STRING} "!format="
  RewriteCond %{REQUEST_URI}  "!/lookup"
  RewriteCond %{REQUEST_URI}  "!/reverse"
  RewriteRule ^([^/]+).php ui/$1.html [R,END]
  # Same but .php suffix is missing.
  RewriteCond %{QUERY_STRING} "!format="
  RewriteCond %{REQUEST_URI}  "!/lookup"
  RewriteCond %{REQUEST_URI}  "!/reverse"
  RewriteRule ^([^/]+) ui/$1.html [R,END]
</Directory>

<Directory "/srv/nominatim/nominatim-ui/dist">
  DirectoryIndex search.html
  Require all granted
</Directory>

Alias /nominatim/ui /srv/nominatim/nominatim-ui/dist

Alias /nominatim /srv/nominatim/nominatim-project/website



I already can access it using a Web browser but I cannot use it as it has the error Error fetching data from http://localhost/nominatimstatus.php?format=json (TypeError: Failed to fetch)

Any tips on what is wrong with the construction of the file or is it totally wrong that I include the codes there? Thank you.

**Reply by mtmail** [reply]
> Error fetching data from http://localhost/nominatimstatus.php?format=json

See https://github.com/osm-search/nominatim-ui/#configuration

You need to create a new file `dist/theme/config.theme.js` and set the `Nominatim_Config.Nominatim_API_Endpoint` value. Try `http://localhost:8080/` or `http://localhost:8080/nominatim/`

**Reply by cormanttech** [reply]
Hi Sir,

I already did this and I only access the Nominatim API with UI using this command

cd nominatim-ui/dist
python3 -m http.server 8765

I tried to change the port to 8080 but this is what it showns:

![image](https://user-images.githubusercontent.com/55179774/179122134-3f0eaa6f-05a3-40d4-bef8-5ec6f77a8a61.png)

I think I successfully connected the Nominatim API and UI, and already can access it using phyton3 8765 port. But I need to forward it to Apache to access it from WWW and that is where I am having a problem. I followed the instructions here https://nominatim.org/release-docs/latest/admin/Setup-Nominatim-UI/ for Apache settings but still no good.

**Reply by cormanttech** [reply]
Hi All,

So my main problem is this, I already set up the Nominatim API and Nominatim UI and connect each other. I can now access it on my localhost browser with the help of this instruction

 
![image](https://user-images.githubusercontent.com/55179774/179169375-8b613efd-5fd5-4024-ba9d-b5d324448788.png)


I just changed the phyton3 to 8080 like this:
python3 -m http.server 8080

And I can finally use the Nominatim OSM and search for places:

 
![image](https://user-images.githubusercontent.com/55179774/179169388-d683a2ae-0dbd-499c-8caf-a29c4c3e5e42.png)


But when I do the Apache forwarding for it to be accessed in the web from the same page of the instruction:
 
![image](https://user-images.githubusercontent.com/55179774/179169406-a93aceaf-ed86-4b46-8ab6-a9ea34681223.png)

It succeeds in deploying it from the web but I cannot search any place and gets this error
![image](https://user-images.githubusercontent.com/55179774/179169426-fdc332ba-760d-4dbf-8d6e-a49e23e7e290.png)

 
I included the directory codes like from the Image above in /etc/apache2/conf-enabled/nominatim.conf

<Directory "/srv/nominatim/nominatim-project/website">
  Options FollowSymLinks MultiViews
  AddType text/html   .php
  DirectoryIndex search.php
  Require all granted

 RewriteEngine On

  # This must correspond to the URL where nominatim can be found.
  RewriteBase "/nominatim/"

  # If no endpoint is given, then use search.
  RewriteRule ^(/|$)   "search.php"

  # If format-html is explicity requested, forward to the UI.
  RewriteCond %{QUERY_STRING} "format=html"
  RewriteRule ^([^/]+).php ui/$1.html [R,END]
  # Same but .php suffix is missing.
  RewriteCond %{QUERY_STRING} "format=html"
  RewriteRule ^([^/]+) ui/$1.html [R,END]

  # If no format parameter is there then forward anything
  # but /reverse and /lookup to the UI.
  RewriteCond %{QUERY_STRING} "!format="
  RewriteCond %{REQUEST_URI}  "!/lookup"
  RewriteCond %{REQUEST_URI}  "!/reverse"
  RewriteRule ^([^/]+).php ui/$1.html [R,END]
  # Same but .php suffix is missing.
  RewriteCond %{QUERY_STRING} "!format="
  RewriteCond %{REQUEST_URI}  "!/lookup"
  RewriteCond %{REQUEST_URI}  "!/reverse"
  RewriteRule ^([^/]+) ui/$1.html [R,END]
</Directory>

<Directory "/srv/nominatim/nominatim-ui/dist">
  DirectoryIndex search.html
  Require all granted
</Directory>

Alias /nominatim/ui /srv/nominatim/nominatim-ui/dist

Alias /nominatim /srv/nominatim/nominatim-project/website


Where I directed the Nominatim API from this instruction:

 
![image](https://user-images.githubusercontent.com/55179774/179169454-0671e894-27a8-4f13-b391-fbcb02e4a4f5.png)


Is there anything wrong that I did or is there something missing? Thank you!

**Comment by tHesTx** [comment]
Thank you for this question, I was trying for few days to get an answer to this...now I know it's normal. My problem is that when I try to reach this page: `http://localhost:8081/search.php?q=avenue%20pasteur`, in browser I get this error:

> This site can’t be reachedThe webpage at http://localhost:8081/search.php?q=avenue%20pasteur might be temporarily down or it may >have moved permanently to a new web address.
>ERR_SOCKET_NOT_CONNECTED

I created this container:

```
docker run -it --shm-size=1g \
  -e PBF_URL=https://download.geofabrik.de/europe/monaco-latest.osm.pbf \
  -e REPLICATION_URL=https://download.geofabrik.de/europe/monaco-updates/ \
  -e IMPORT_WIKIPEDIA=false \
  -e NOMINATIM_PASSWORD=very_secure_password \
  -v nominatim-data:/var/lib/postgresql/12/main \
  -p 8081:8081 \
  --name nominatim \
  mediagis/nominatim:4.0
```

**Reply by leonardehrenfried** [reply]
Why are you using port 8081 and not 8080?

**Reply by tHesTx** [reply]
I have other container on 8080

**Reply by tHesTx** [reply]
I've tried on 8080 but same error

**Reply by tHesTx** [reply]
entire log here:

```
 $ docker run -it --shm-size=1g \
  -e PBF_URL=https://download.geofabrik.de/europe/monaco-latest.osm.pbf \
  -e REPLICATION_URL=https://download.geofabrik.de/europe/monaco-updates/ \
  -e IMPORT_WIKIPEDIA=false \
  -e NOMINATIM_PASSWORD=password \
  -v nominatim-data:/var/lib/postgresql/12/main \
  -p 8081:8081 \
  --name nominatim \
  mediagis/nominatim:4.0
+ trap stopServices TERM
+ /app/config.sh
+ id nominatim
+ useradd -m -p password  nominatim
+ IMPORT_FINISHED=/var/lib/postgresql/12/main/import-finished
+ TOKENIZER_DIR=/nominatim/tokenizer
+ '[' '!' -f /var/lib/postgresql/12/main/import-finished ']'
+ chown -R nominatim:nominatim /nominatim
+ '[' '!' -d /nominatim/tokenizer ']'
+ echo 'No tokenizer configuration found. Copying from persistent volume into project directory.'
No tokenizer configuration found. Copying from persistent volume into project directory.
+ cp -r /var/lib/postgresql/12/main/tokenizer /nominatim/tokenizer
+ chown -R nominatim:nominatim /nominatim
+ service postgresql start
 * Starting PostgreSQL 12 database server                                                                                                                                                    [ OK ] 
+ cd /nominatim
+ sudo -E -u nominatim nominatim refresh --website --functions
2022-07-22 07:55:40: Using project directory: /nominatim
2022-07-22 07:55:40: Create functions
2022-07-22 07:55:43: Setting up website directory at /nominatim/website
+ service apache2 start
 * Starting Apache httpd web server apache2                                                                                                                                                         AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 172.17.0.2. Set the 'ServerName' directive globally to suppress this message
 * 
+ wait
+ tail -f /var/log/postgresql/postgresql-12-main.log
2022-07-22 07:55:35.270 UTC [46] LOG:  starting PostgreSQL 12.11 (Ubuntu 12.11-0ubuntu0.20.04.1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0, 64-bit
2022-07-22 07:55:35.276 UTC [46] LOG:  listening on IPv4 address "0.0.0.0", port 5432
2022-07-22 07:55:35.276 UTC [46] LOG:  listening on IPv6 address "::", port 5432
2022-07-22 07:55:35.277 UTC [46] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
2022-07-22 07:55:35.747 UTC [47] LOG:  database system was interrupted; last known up at 2022-07-21 13:03:07 UTC
2022-07-22 07:55:35.773 UTC [47] LOG:  database system was not properly shut down; automatic recovery in progress
2022-07-22 07:55:35.794 UTC [47] LOG:  redo starts at 1/A8A8D550
2022-07-22 07:55:36.286 UTC [47] LOG:  invalid record length at 1/A8AC8B38: wanted 24, got 0
2022-07-22 07:55:36.286 UTC [47] LOG:  redo done at 1/A8AC8B00
2022-07-22 07:55:36.347 UTC [46] LOG:  database system is ready to accept connections
```

**Reply by leonardehrenfried** [reply]
Once more, your are connecting to port 8081 inside the container but nothing is listening on it. That isn't going to work.

Try, 

```
-p 8081:8080
```

**Reply by tHesTx** [reply]
That was it...thank you so much, beginner...
